### ISSUE TITLE
feature request: no reject literal

### ISSUE THREAD

**Issue by jonathanong** [eslint/eslint#4092]
similar to http://eslint.org/docs/rules/no-throw-literal, but for promises

i'd like to throw a lint error when you reject a literal with promises:

``` js
Promise.reject('an error happened!')
```

but not if it's an actual error:

``` js
Promise.reject(new Error('an error happened'))
```

it could be a new rule or it could be an option for `no-throw-literal`

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
I'm not aware of any guidance suggesting that rejecting a literal is bad practice. What's the rationale?

**Comment by jonathanong** [comment]
you would have to ask someone who knows ES better than me. it's usually recommended, ex: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/reject#Description and `Promise.reject()` is the promise-version of `throw`. throwing an error in a promise returns a promise with the error as the rejection reason. 

also, if you throw a non-error in a promise and never handle it, then the unhandled error reason would be not be useful at all (similar to throwing non-errors)

**Comment by nzakas** [comment]
That's the first place I've seen recommend using an error, so I'm not ready to create a core rule for it. I'd suggest creating a custom rule for now. When ES6 best practices become more obvious, we can revisit.

**Comment by TheSavior** [comment]
We just ran into this. Let me provide an example to demonstrate the issue with rejecting a non error.

![screen shot 2016-12-01 at 2 06 14 pm](https://cloud.githubusercontent.com/assets/249164/20814628/5faeca4e-b7cf-11e6-8592-132b944f6ccc.png)

```
window.onunhandledrejection = function(e) { window.last = e; };
(e) { window.last = e; }
Promise.reject(new Error('foo'))
Promise {[[PromiseStatus]]: "rejected", [[PromiseValue]]: Error: foo
    at <anonymous>:1:16}
VM1739:1 Uncaught (in promise) Error: foo(…)(anonymous function) @ VM1739:1
window.last.reason
Error: foo(…)
window.last.reason.stack
"Error: foo
    at <anonymous>:1:16"
Promise.reject('foo')
Promise {[[PromiseStatus]]: "rejected", [[PromiseValue]]: "foo"}
VM1895:1 Uncaught (in promise) foo(anonymous function) @ VM1895:1
window.last.reason
"foo"
window.last.reason.stack
undefined
Promise.reject()
Promise {[[PromiseStatus]]: "rejected", [[PromiseValue]]: undefined}
VM1930:1 Uncaught (in promise) undefined(anonymous function) @ VM1930:1
window.last.reason
undefined
```

**Comment by not-an-aardvark** [comment]
> When ES6 best practices become more obvious, we can revisit.

It's been over a year, so I think we can probably revisit this now.

I'm willing to champion this proposal; it's useful to ensure that async errors have stack traces. We have [`no-throw-literal`](http://eslint.org/docs/rules/no-throw-literal) which does basically the same thing with synchronous errors. Bluebird, a very popular Promise library, [emits a warning](http://bluebirdjs.com/docs/warning-explanations.html#warning-a-promise-was-rejected-with-a-non-error) by default when the user does this.

**Comment by platinumazure** [comment]
Probably good to get a formal proposal up, but this has my :+1:.

**Comment by not-an-aardvark** [comment]
Good idea. I think I'll close this and repost a new issue with the proposal for better visibility.

**Comment by not-an-aardvark** [comment]
https://github.com/eslint/eslint/issues/7685
